src/ngraph/node.hpp

```diff
@@ -139,6 +148,41 @@ namespace ngraph
     public:
         virtual ~Node();
 
+        /// Tests if a node is of op type T
+        template <typename NodeType>
+        bool has_type() const
+        {
+            return &get_type_info() == &NodeType::type_info;
+        }
+
+        /// Casts a Node to a shared_ptr<T> if is of type T, nullptr otherwise;
+        template <typename NodeType>
+        std::shared_ptr<NodeType> as_type()
```

Comment:
I don't like that this further burns in the requirement that everything in every class in ngraph must be implemented as a shared_ptr. Can't you just return a NodeType? Or get rid of this method?